src/core/sock.c

```diff
@@ -188,31 +194,54 @@ void nn_sock_zombify (struct nn_sock *self)
     nn_ctx_leave (&self->ctx);
 }
 
+/*  Stop the socket.  This will prevent new calls from aquiring a
+    hold on the socket, cause endpoints to shut down, and wake any
+    threads waiting to recv or send data. */
+void nn_sock_stop (struct nn_sock *self)
+{
+    nn_ctx_enter (&self->ctx);
+    nn_fsm_stop (&self->fsm);
+    nn_ctx_leave (&self->ctx);
+}
+
 int nn_sock_term (struct nn_sock *self)
 {
     int rc;
     int i;
 
-    /*  Ask the state machine to start closing the socket. */
-    nn_ctx_enter (&self->ctx);
-    nn_fsm_stop (&self->fsm);
-    nn_ctx_leave (&self->ctx);
+    /*  NOTE: nn_sock_stop must have already been called. */
 
-    /*  Shutdown process was already started but some endpoints may still
-        alive. Here we are going to wait till they are all closed. */
-    rc = nn_sem_wait (&self->termsem);
-    if (nn_slow (rc == -EINTR))
-        return -EINTR;
-    errnum_assert (rc == 0, -rc);
+    /*  Some endpoints may still be alive.  Here we are going to wait
+        till they are all closed.  This loop is not interruptible, because
+        making it so would leave a partially cleaned up socket, and we don't
+        have a way to defer resource deallocation. */
+    for (;;) {
+        rc = nn_sem_wait (&self->termsem);
+        if (nn_slow (rc == -EINTR))
+            continue;
+        errnum_assert (rc == 0, -rc);
+        break;
+    }
+
+    /*  Also, wait for all holds on the socket to be released.  */
+    for (;;) {
+        rc = nn_sem_wait (&self->relesem);
+        if (nn_slow (rc == -EINTR))
+            continue;
+        errnum_assert (rc == 0, -rc);
+        break;
+    }
```

Comment:
Same question; how long could this take? Is it denying service to other sockets in the meantime? (i.e., by holding the global mutex? or is the global mutex not held here?)